instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb

```diff
@@ -191,11 +193,11 @@ def request_span_attributes(env)
             attributes
           end
 
-          def detach_contexts(request)
-            request.env[TOKENS_KEY]&.reverse_each do |token|
-              OpenTelemetry::Context.detach(token)
-              OpenTelemetry::Trace.current_span.finish
-            end
+          def detach_context(request)
+            return nil unless request.env[TOKEN_KEY]
+
+            OpenTelemetry::Trace.current_span.finish
```

Comment:
The lack of symmetry (with `on_start`'s use of `Foo.context_with_span`) threw me quite a bit here. Nothing to be done - just noting my confusion. Basically down to `on_start` managing contexts explicitly, whereas here we use the implicit "current context".

Reply:
I wonder if it would make sense to store a reference to the rack span in the rack request env so we can close it later, rather than relying on in the implicit `current_span`. 

With the current approach we're vulnerable to a scenario where some misconfigured or buggy instrumentation opening a span, setting the context, and not closing it... right? It's probably not likely, but still plausible.

Reply:
If the Rack span is not the current span at this point then it means the context token is out of sync also and we will likely get a detach error. 

If this tracked the token and the `Context` together as a tuple then it would fetch the Rack span from it instead and ensure it is closing the right span and the token/Context would be the same.

If something went wrong with a different implicit "current" context not being `detached`, it would still result in a detach error but at least the Rack span would be the one we are trying to finish. 
e.g. 

```ruby
return nil unless request.env[RACK_OTEL_CONTEXT_KEY]

[token, rack_ctx] = request.env[RACK_OTEL_CONTEXT_KEY]

OpenTelemetry::Trace.current_span(rack_ctx).finish
OpenTelemetry::Context.detach(token)
```

Reply:
I think that makes sense to me, see https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/965/commits/0b666b995053662ad8c72a462478f6defa20a75c